main_boaz.tex

```diff
@@ -0,0 +1,967 @@
+
+
+Just as the set of all size-$n$ real vectors is written $\mathbb{R}^n$, and the set of all $m\times n$ real matrices is often written $\mathbb{R}^{m \times n}$ (which makes sense because one sometimes writes $Y^X$ for the set of all functions from $X$ to $Y$), we write $\mathbb{R}^{\nset{foo}{..2}, \nset{bar}{..3}}$ for the set of all tensors with shape $\tuple{\nset{foo}{..2}, \nset{bar}{..3}}$.
+We write $\mathbb{R}^{\name{foo},\name{bar}}$ for the set of all tensors whose shape has the form $\{ \name{foo}[S], \name{bar}[T] \}$ for some subsets $S,T$ of the natural numbers.
+
+
+We also allow \emph{partial indexing}:
+\begin{align*}
+A_{\nidx{foo}{1}} &= \nmatrix{}{bar}{
+  3 & 1 & 4
+}
+\\
+A_{\nidx{bar}{3}} &= \nmatrix{}{foo}{
+  4 & 9
+}.
+\end{align*}
+
+In general, if $A$ is a tensor of shape $\mathcal{S}$ where $d=|\mathcal{S}|$, and $i_1,\ldots,i_\ell$ are elements in distinct slices $S_1,\ldots,S_\ell \in \mathcal{S}$ then $A(i_1,\ldots,i_\ell)$ is the tensor of shape $\mathcal{S} \setminus \{ S_1,\ldots, S_\ell \}$ that maps $j_1,\ldots,j_d-{\ell}$ into the $A(i_1,\ldots,i_\ell,j_1,\ldots,j_{d-\ell})$.
+
+
+
+\paragraph{Proposed convention:} We propose that if $i$ is an axis belonging to an axis  \emph{not} in  $type(A)$ then $A(i)=A$.  This can be convenient for example to use a $\{ \name{width} ,\name{height} \}$ type tensor in a function that expects a $\{ \name{width} , \name{height} , \name{batch} \}$ tensor by essentially treating the former as if it has a trivial (only one coordinate) $\name{batch}$ axis.
+
+
+\subsection{Named tensor operations}
+\label{sec:operations}
+
+\subsection{General operations}
+
+An \emph{operator}  takes as input one or more tensors and outputs as input a tensor. For example:
+
+$CONV: \mathbb{R}^{\name{width},\name{height}} \times \mathbb{R}^{\name{width},\name{height}} \rightarrow \mathbb{R}^{\name{width},\name{height}}$
+
+is the two dimensional convolution defined as following: $CONV(W,X)=Y$ of shape $\{ \name{width}(X),\name{height}(X) \}$ such that for every $(i,j) \in (\name{width},\name{height})$, 
+
+$Y_{x,y} = \sum_{(a,b) \in (\name{width}(W),\name{height}(W)} X_{x+a,y+b} W_{a,b}$
+
+Another way to write it is that 
+
+$Y_{x,y} =  X_{\name{width}[x-\ell:x+\ell],\name{height}[y-\ell,y+\ell]} \cdot W$ where $|\name{width}(W)|=|\name{height}(W)|=2\ell+1$.
+
+(We are thinking of the axes of $W$ here as indexed by integers $\{ -\ell, -\ell+1, \ldots, 0, 1,\ldots, \ell \}$. We assume here that any ``out of bound'' value defaults to zero.)
+
+\paragraph{Dangling and aligned axes} We use the following conventions for an operator $F$ taking one or more tensors $X_1,X_2,\ldots,X_\ell$ when these tensors contain axes that do not appear in the signature of $F$:
```

Comment:
Do you also need to spell out when automatic casting can occur for this line? For example, if

A ∈ R^{ax^1}
F : R^{ax} -> R

can F be applied to A? (I hope no.)

Reply:
Yes F can be applied to A. In fact, that is probably the main reason to use annotation. (If you don't want to use inheritence that ax^1 automatically is cast to ax then there is no need to use annotation - you can just use different names.)

Reply:
OK, so the rules would need to say that automatic casting can occur when applying an operator to a tensor too. Another test case:

A ∈ R^{ax^2, ax^3}
F : R^{ax^1, ax^2} -> R

It seems like F(A) could have two interpretations depending on the order in which the indices are resolved. If I try to resolve the first index first, I would see that A has two ax indices, and fail.

But if I try to resolve the second index first, I would align A's ax^2 to F's ax^2, and then would I cast A's ax^3 to F's ax^1, or would it fail?

Reply:
I think this is a case where the most specific match is to align  2 to 2 and 3 to 1. However, it's probably a case where the author should be explicit.

However, this rule might not be a bad thing. Suppose that you have three layers  layer^in , layer^mid, layer^out 

and an F that maps layer^in to layer^out 

Now if you apply F to a tensor that has the shape { layer^in , layer^mid } it actually makes sense that it would map in to in and mid to out.